stock_barcodes/wizard/stock_barcodes_read_picking.py

```diff
@@ -0,0 +1,293 @@
+                    qty_need = line.product_uom_qty - line.qty_done
+                    if self.product_qty < qty_need:
+                        qty_assigned = self.product_qty
+                    else:
+                        qty_assigned = qty_need
+                    line.qty_done += qty_assigned
+                else:
+                    line.qty_done += self.product_qty - qty_assigned
+            self.update_quantity_done(lines)
+        else:
+            # Create an extra stock move line if this product has an
+            # initial demand.
+            moves = self.picking_id.move_lines.filtered(lambda m: (
+                m.product_id == self.product_id and m.state == 'assigned'))
+            if not moves:
+                self._set_messagge_info(
+                    'info',
+                    _('There are no stock moves to assign this operation'))
+                return False
+            else:
+                vals = self._prepare_move_line_values(moves[0])
+                move_line = StockMoveLine.create(vals)
+            self.update_quantity_done(move_line)
+        return True
+
+    def update_quantity_done(self, move_lines):
+        self.picking_product_qty = sum(
+            move_lines.mapped('move_id.quantity_done'))
+
+    def select_cadidate_picking(self, picking_candidates):
+        self._set_candidate_pickings(picking_candidates)
+
+    def _candidate_picking_selected(self):
+        if len(self.candidate_picking_ids) == 1:
+            return self.candidate_picking_ids.picking_id
+        else:
+            return self.env['stock.picking'].browse()
+
+    def check_done_conditions(self):
+        res = super().check_done_conditions()
+        if self.product_id.tracking != 'none' and not self.lot_id:
+            self._set_messagge_info('info', _('Waiting for input lot'))
+            return False
+        if not self.picking_id:
+            if not self._search_candidate_pickings():
+                self._set_messagge_info('info', _('Select one picking'))
+                return False
+        if self.picking_id != self._candidate_picking_selected():
+            self._set_messagge_info('info', _('Select one picking'))
+            return False
+        return res
+
+    def _prepare_move_line_undo_domain(self, log_scan=False):
+        return [
+            ('picking_id', '=', log_scan.picking_id.id),
+            ('product_id', '=', log_scan.product_id.id),
+            ('lot_id', '=', log_scan.lot_id.id),
+            ('move_id.picking_id.picking_type_id.code', '=',
+             self.env.context.get('default_picking_type_code')),
+            ('state', '=', 'assigned'),
+        ]
+
+    def action_undo_last_scan(self):
+        res = super().action_undo_last_scan()
+        log_scan = first(self.scan_log_ids.filtered(
+            lambda x: x.create_uid == self.env.user))
+        if log_scan:
+            domain = self._prepare_move_line_undo_domain(log_scan=log_scan)
```

Comment:
Not sure if this is going to return always your last scan or others. The only way to be sure is to add a scan_log_id at sml level.